### PR TITLE
When using synctl with workers, don't start the main synapse automatically

### PR DESCRIPTION
**Pull request by erikjohnston**

### CODE REVIEW

File: synapse/app/synctl.py [hunk 1]
```diff
@@ -200,14 +203,29 @@ def main():
...
+                worker_pidfile,
+                worker_config.get("synctl_cache_factor") or cache_factor,
+            ))
+        else:
+            worker_pidfile = worker_config["worker_pid_file"]
+            worker_daemonize = worker_config["worker_daemonize"]
+            assert worker_daemonize, "In config %r: expected '%s' to be True" % (
+                worker_configfile, "worker_daemonize")
+            worker_cache_factor = worker_config.get("synctl_cache_factor")
+            workers.append(Worker(
```
Comment (richvdh) [inline synapse/app/synctl.py]: we can share this between the two branches provided you initialise `worker_cache_factor` correctly.

**Comment by erikjohnston** [comment]
Right, sorry, I managed to mess up testing this. Two sticking points:
1. `synapse.app.homeserver` doesn't look at `worker_pidfile` etc
2. There's a few places where we assume if `worker_app` is `None` then we're a worker.

For 1 I've made synctl complain if it sees `worker_` config and for 2 I've canonicalised worker_app to be None if its the main process

**Comment by richvdh** [comment]
I can't help thinking it would be better if `synapse.app.homeserver` *did* look at the `worker_*` options?

**Comment by erikjohnston** [comment]
> I can't help thinking it would be better if `synapse.app.homeserver` *did* look at the `worker_*` options?

Yes, or even to not have the duplication between `worker_pidfile` and `pidfile` params, but I think that ends up with quite a bit of synapse fiddling (which I'm not enthusiastic about dong right now)

**Pull request merged**